libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -192,11 +223,118 @@ bool AP_OABendyRuler::update(const Location& current_loc, const Location& destin
+                    float bearing_test2 = wrap_180(test_vertical_bearings_step2[j]);
+                    Location test_loc2 = test_loc;
+                    test_loc2.offset_bearing_and_pitch(bearing_to_dest2,bearing_test2 ,distance2);
+
+                    // calculate minimum margin to fence and obstacles for this scenario
+                    float margin2 = calc_avoidance_margin(test_loc, test_loc2);
+                    if (margin2 > _margin_max) {
+                        // all good, now project in the chosen direction by the full distance
+                        destination_new = current_loc;
+                        destination_new.offset_bearing_and_pitch(bearing_to_dest,vertical_bearing_delta, distance_to_dest);
+                        _current_lookahead = MIN(_lookahead, _current_lookahead * 1.1f);
+                        // if the chosen direction is directly towards the destination turn off avoidance
+                        // i == 0 && j == 0 implies no deviation from bearing to destination
+                        const bool active = (i != 0 || j != 0);
+                        AP::logger().Write_OABendyRuler((uint8_t)OABendyType::OA_BENDY_VERTICAL, active, bearing_to_dest, vertical_bearing_delta, false ,margin, destination, destination_new);
+                        return active;
+                    }
+                }
+            }
+
+        }        
+    }   
+
+    float chosen_vertical_bearing;
+    if (have_best_vertical_bearing) {
+        // none of the directions tested were OK for 2-step checks. Choose the direction
+        // that was best for the first step
+        chosen_vertical_bearing = best_vertical_bearing;
+        _current_lookahead = MIN(_lookahead, _current_lookahead * 1.05f);
+    } else {
+        // none of the possible paths had a positive margin. Choose
+        // the one with the highest margin
+        chosen_vertical_bearing = best_margin_pitch;
+        _current_lookahead = MAX(_lookahead * 0.5f, _current_lookahead * 0.9f);
+    }
+
+    // calculate new target based on best effort
+    destination_new = current_loc;
+    destination_new.offset_bearing_and_pitch(bearing_to_dest,chosen_vertical_bearing, distance_to_dest);
+
+    // log results
+    AP::logger().Write_OABendyRuler((uint8_t)OABendyType::OA_BENDY_VERTICAL, true, bearing_to_dest, chosen_vertical_bearing,false, best_margin, destination, destination_new);
 
     return true;
 }
 
+AP_OABendyRuler::OABendyType AP_OABendyRuler::get_type() const
+{
+    switch (_bendy_type) {
+        case (uint8_t)OABendyType::OA_BENDY_VERTICAL:
+        #if APM_BUILD_TYPE(APM_BUILD_ArduCopter)
+            return OABendyType::OA_BENDY_VERTICAL;
+            break;
+        #endif
+
+        case (uint8_t)OABendyType::OA_BENDY_HORIZONTAL:
+        default:
+            return OABendyType::OA_BENDY_HORIZONTAL;
+            break;
+    }
+    return OABendyType::OA_BENDY_HORIZONTAL;
```

Comment:
this last line can never be reached I think.  Again it does no harm though and perhaps the compiler will complain without it.